src/jit/compiler.cpp

```diff
@@ -516,63 +488,475 @@ var_types    Compiler::argOrReturnTypeForStruct(unsigned size, CORINFO_CLASS_HAN
+                howToPassStruct = SPK_ByValue;
+                useType = TYP_UNKNOWN;
+
+#else  //  _TARGET_XXX_
+
+                noway_assert(!"Unhandled TARGET in getArgTypeForStruct (with FEATURE_MULTIREG_ARGS=1)");
+
+#endif  //  _TARGET_XXX_
+
             }
         }
-#endif // FEATURE_MULTIREG_RET
+        else  // (structSize > MAX_PASS_MULTIREG_BYTES)
+        {
+            // We have a (large) struct that can't be replaced with a "primitive" type
+            // and can't be passed in multiple registers
 
-#if FEATURE_MULTIREG_ARGS
-        if (!forReturn)
+#if defined(_TARGET_X86_) || defined(_TARGET_ARM_)
+
+            // Otherwise we pass this struct by value on the stack 
+            // setup wbPassType and useType indicate that this is passed by value according to the X86/ARM32 ABI
+            howToPassStruct = SPK_ByValue;
+            useType = TYP_UNKNOWN;
+
+#elif defined(_TARGET_AMD64_) || defined(_TARGET_ARM64_)
+
+            // Otherwise we pass this struct by reference to a copy
+            // setup wbPassType and useType indicate that this is passed using one register (by reference to a copy)
+            howToPassStruct = SPK_ByReference;
+            useType = TYP_UNKNOWN;
+
+#else  //  _TARGET_XXX_
+
+            noway_assert(!"Unhandled TARGET in getArgTypeForStruct");
+
+#endif  //  _TARGET_XXX_
+
+        }
+    }
+
+    // 'howToPassStruct' must be set to one of the valid values before we return
+    assert(howToPassStruct != SPK_Unknown);
+    if (wbPassStruct != nullptr)
+    {
+        *wbPassStruct = howToPassStruct;
+    }
+    return useType;
+}
+
+//-----------------------------------------------------------------------------
+// getReturnTypeForStruct: 
+//     Get the type that is used to return values of the given struct type.
+//     If you have already retrieved the struct size then it should be 
+//     passed as the optional third argument, as this allows us to avoid
+//     an extra call to getClassSize(clsHnd)
+//
+// Arguments:
+//    clsHnd         - the handle for the struct type
+//    wbReturnStruct - An "out" argument with information about how
+//                     the struct is to be returned
+//    structSize     - the size of the struct type, 
+//                     or zero if we should call getClassSize(clsHnd)
+//
+// Return Value:
+//    For wbReturnStruct you can pass a 'nullptr' and nothing will be written
+//     or returned for that out parameter.
+//    When *wbReturnStruct is SPK_PrimitiveType this method's return value
+//       is the primitive type used to return the struct.
+//    When *wbReturnStruct is SPK_ByReference this method's return value
+//       is TYP_UNKNOWN when the struct type is returned using a return buffer
+//    When *wbReturnStruct is SPK_ByValue or SPK_ByValueAsHfa this method's return value
+//       is TYP_STRUCT when the struct type is returned using multiple registers.
```

Comment:
These comments don't appear to cover all the cases - if these cases can't occur, the comment make it clear:
I think it's the case that when *wbReturnStruct is SPK_ByReference, the return value is always TYP_UNKNOWN, AND (not "when") the struct type is (always) returned using a return buffer.
I also think it's the case that when *wbReturnStruct is SPK_ByValue or SPK_ByValueAsHfa, then the methods return value is always TYP_STRUCT, AND (not "when") the struct type is returned using multiple registers.